File: packages/cortex-python/cortex_sdk/resources/audio.py
"""Audio resource."""

from __future__ import annotations

from typing import Any, BinaryIO, Dict, Optional, Union

from cortex_sdk._http import AsyncHTTPClient, SyncHTTPClient
from cortex_sdk.types import AudioTranscription


class Audio:
    """Synchronous audio resource."""

    def __init__(self, http: SyncHTTPClient, *, default_pool: Optional[str] = None) -> None:
        self._http = http
        self._default_pool = default_pool

    def transcribe(
        self,
        *,
        file: Union[BinaryIO, bytes],
        model: Optional[str] = None,
        language: Optional[str] = None,
        prompt: Optional[str] = None,
        response_format: Optional[str] = None,
        temperature: Optional[float] = None,
        pool: Optional[str] = None,
    ) -> AudioTranscription:
        files: dict[str, Any] = {"file": file}
        form_data: dict[str, Any] = {}
        if model is not None:
            form_data["model"] = model
        if language is not None:
            form_data["language"] = language
        if prompt is not None:
            form_data["prompt"] = prompt
        if response_format is not None:
            form_data["response_format"] = response_format
        if temperature is not None:
            form_data["temperature"] = str(temperature)

        # Resolve pool: per-request > client default > none
        resolved_pool = pool or self._default_pool
        extra_headers: Optional[Dict[str, str]] = None
        if resolved_pool:
            extra_headers = {"x-cortex-pool": resolved_pool}

        data = self._http.request(
            "POST",
            "/v1/audio/transcriptions",
            files=files,
            content=form_data,
            extra_headers=extra_headers,
        )
        return AudioTranscription.model_validate(data)


class AsyncAudio:
    """Asynchronous audio resource."""

    def __init__(self, http: AsyncHTTPClient, *, default_pool: Optional[str] = None) -> None:
        self._http = http
        self._default_pool = default_pool

    async def transcribe(
        self,
        *,
        file: Union[BinaryIO, bytes],
        model: Optional[str] = None,
        language: Optional[str] = None,
        prompt: Optional[str] = None,
        response_format: Optional[str] = None,
        temperature: Optional[float] = None,
        pool: Optional[str] = None,
    ) -> AudioTranscription:
        files: dict[str, Any] = {"file": file}
        form_data: dict[str, Any] = {}
        if model is not None:
            form_data["model"] = model
        if language is not None:
            form_data["language"] = language
        if prompt is not None:
            form_data["prompt"] = prompt
        if response_format is not None:
            form_data["response_format"] = response_format
        if temperature is not None:
            form_data["temperature"] = str(temperature)

        # Resolve pool: per-request > client default > none
        resolved_pool = pool or self._default_pool
        extra_headers: Optional[Dict[str, str]] = None
        if resolved_pool:
            extra_headers = {"x-cortex-pool": resolved_pool}

        data = await self._http.request(
            "POST",
            "/v1/audio/transcriptions",
            files=files,
            content=form_data,
            extra_headers=extra_headers,
        )
        return AudioTranscription.model_validate(data)

File: packages/cortex-dart/lib/src/resources/audio.dart
/// Audio resource.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:http/http.dart' as http;

import '../errors.dart';
import '../http_client.dart';
import '../types.dart';

/// Provides access to audio endpoints.
class AudioResource {
  final CortexHttpClient _client;
  final String _baseUrl;
  final String? _defaultPool;

  /// Creates an [AudioResource].
  AudioResource({
    required CortexHttpClient client,
    required String baseUrl,
    String? defaultPool,
  })  : _client = client,
        _baseUrl = baseUrl,
        _defaultPool = defaultPool;

  /// Transcribes an audio file.
  ///
  /// [file] is the audio file bytes.
  /// [fileName] is the file name (e.g., "audio.mp3").
  /// [model] is the transcription model to use. If omitted, the pool's default model is used.
  /// [language] is an optional language hint.
  /// [pool] overrides the client-level default pool for this request.
  Future<AudioTranscription> transcribe({
    required Uint8List file,
    required String fileName,
    String? model,
    String? language,
    String? pool,
  }) async {
    if (fileName.isEmpty) {
      throw const CortexValidationException(
        message: 'File name must not be empty.',
        parameter: 'fileName',
      );
    }
    if (model != null && model.isEmpty) {
      throw const CortexValidationException(
        message: 'Model must not be empty.',
        parameter: 'model',
      );
    }

    // Resolve pool: per-request > client default > none
    final resolvedPool = pool ?? _defaultPool;
    final extraHeaders = <String, String>{};
    if (resolvedPool != null) {
      extraHeaders['x-cortex-pool'] = resolvedPool;
    }

    // For multipart form data, we use the HTTP client's post method
    // with a JSON body that includes the base64-encoded file.
    final json = await _client.post(
      '$_baseUrl/v1/audio/transcriptions',
      body: {
        'file': base64Encode(file),
        'fileName': fileName,
        if (model != null) 'model': model,
        if (language != null) 'language': language,
      },
      extraHeaders: extraHeaders.isNotEmpty ? extraHeaders : null,
    );
    return AudioTranscription.fromJson(json);
  }
}
